pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -941,6 +951,7 @@ func (m *kubeGenericRuntimeManager) computePodActions(ctx context.Context, pod *
 			message = fmt.Sprintf("Container %s failed startup probe", container.Name)
 			reason = reasonStartupProbe
 		} else if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) &&
+			!types.IsStaticPod(pod) &&
```

Comment:
it's _really_ hard to understand how this carve-out for static pods combines with ignoring resource changes and skipping the `restart = true` case above when InPlacePodVerticalScaling is enabled

At first glance, it looks like an update to a static pod manifest's resources which previously would have triggered a restart might now fall through to this case, get ignored because it's a static pod, and the container will be kept as is. Is that possible?

Reply:
From observation, on manifest update (save), the pod remove/add happens much earlier.
```bash
  2473:	// HandlePodRemoves is the callback in the SyncHandler interface for pods
  2474:	// being removed from a config source.
=>2475:	func (kl *Kubelet) HandlePodRemoves(pods []*v1.Pod) {
  2476:		start := kl.clock.Now()
  2477:		for _, pod := range pods {
  2478:			kl.podManager.DeletePod(pod)
  2479:			if kubetypes.IsMirrorPod(pod) {
  2480:				kl.handleMirrorPod(pod, start)
(dlv) bt
0  0x0000000003ad70c0 in k8s.io/kubernetes/pkg/kubelet.(*Kubelet).HandlePodRemoves
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2475
1  0x0000000003ad582c in k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoopIteration
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2268
2  0x0000000003ad3fdc in k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoop
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2205
3  0x0000000003acb5c8 in k8s.io/kubernetes/pkg/kubelet.(*Kubelet).Run
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1560
4  0x0000000003b365fc in k8s.io/kubernetes/cmd/kubelet/app.startKubelet.func1
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:1190
5  0x00000000000803d4 in runtime.goexit
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/.gimme/versions/go1.20.2.linux.arm64/src/runtime/asm_arm64.s:1172
(dlv) c
> k8s.io/kubernetes/pkg/kubelet.(*Kubelet).HandlePodAdditions() /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2393 (hits goroutine(211):1 total:1) (PC: 0x3ad6640)
  2388:		}
  2389:	}
  2390:	
  2391:	// HandlePodAdditions is the callback in SyncHandler for pods being added from
  2392:	// a config source.
=>2393:	func (kl *Kubelet) HandlePodAdditions(pods []*v1.Pod) {
  2394:		start := kl.clock.Now()
  2395:		sort.Sort(sliceutils.PodsByCreationTime(pods))
  2396:		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
  2397:			kl.podResizeMutex.Lock()
  2398:			defer kl.podResizeMutex.Unlock()
(dlv) bt
0  0x0000000003ad6640 in k8s.io/kubernetes/pkg/kubelet.(*Kubelet).HandlePodAdditions
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2393
1  0x0000000003ad5490 in k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoopIteration
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2262
2  0x0000000003ad3fdc in k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoop
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2205
3  0x0000000003acb5c8 in k8s.io/kubernetes/pkg/kubelet.(*Kubelet).Run
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1560
4  0x0000000003b365fc in k8s.io/kubernetes/cmd/kubelet/app.startKubelet.func1
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:1190
5  0x00000000000803d4 in runtime.goexit
   at /root/go/src/k8s.io/kubernetes-kubeletonly-crash/_output/local/.gimme/versions/go1.20.2.linux.arm64/src/runtime/asm_arm64.s:1172
(dlv) c
```

Reply:
I still think the `if` check on line 938 is incorrect if we're going to let static pods skip restart there because in-place is enabled, but exclude them from being handled here

If we're going to exclude static pods, can we do something like this:

```go
func canBeVerticallyScaled(pod *api.Pod) bool {
  if !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
    return false
  }
  if types.IsStaticPod(pod) {
    // TODO: ...
    return false
  }
  return true
}
```

and then use that in all three of these places:
```go
		if _, _, changed := containerChanged(&container, containerStatus); changed &&
			(!canBeVerticallyScaled(pod) || kubecontainer.HashContainerWithoutResources(&container) != containerStatus.HashWithoutResources) {
...
		} else if canBeVerticallyScaled(pod) && !m.computePodResizeAction(pod, idx, containerStatus, &changes) {
...
	// Step 7: For containers in podContainerChanges.ContainersToUpdate[CPU,Memory] list, invoke UpdateContainerResources
	if canBeVerticallyScaled(pod) {
...
```

Reply:
@vinaykul can you respond @liggitt's above comment?

Reply:
I just had a call with Clayton and he reconfirmed that static pods should not be updated via API. So excluding static pods is the correct thing to do. Let me take a closer look at your suggested change to see if I can fold the feature gate into such a function as it looks cleaner.

Reply:
@liggitt I pushed a commit based on your suggested changes, it looks more obvious and cleaner now. PTAL.

Validating static pods with this PR:

  - Feature gate enabled, normal local cluster.
```bash
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /var/run/kubernetes/static-pods/3Bpod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: bpod3
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: ctr
    image: skiibum/testpod:qos
    imagePullPolicy: IfNotPresent
    command: ["tail", "-f", "/dev/null"]
    resources:
      requests:
        cpu: "100m"
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .spec.containers[0].resources.requests
{
  "cpu": "100m"
}
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .status.containerStatuses[0].resourcesAllocated
{
  "cpu": "100m"
}
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .status.containerStatuses[0].resources
{
  "requests": {
    "cpu": "100m"
  }
}
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# 
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /var/run/kubernetes/static-pods/3Bpod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: bpod3
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: ctr
    image: skiibum/testpod:qos
    imagePullPolicy: IfNotPresent
    command: ["tail", "-f", "/dev/null"]
    resources:
      requests:
        cpu: "200m"
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .spec.containers[0].resources
{
  "requests": {
    "cpu": "200m"
  }
}
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .status.containerStatuses[0].resourcesAllocated
{
  "cpu": "200m"
}
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .status.containerStatuses[0].resources
{
  "requests": {
    "cpu": "200m"
  }
}
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash#
```

  - Feature gate disabled, normal local cluster.
```bash
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /var/run/kubernetes/static-pods/3Bpod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: bpod3
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: ctr
    image: skiibum/testpod:qos
    imagePullPolicy: IfNotPresent
    command: ["tail", "-f", "/dev/null"]
    resources:
      requests:
        cpu: "200m"
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# 
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .spec.containers[0].resources
{
  "requests": {
    "cpu": "200m"
  }
}
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .status.containerStatuses[0].resourcesAllocated
null
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .status.containerStatuses[0].resources
null
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /var/run/kubernetes/static-pods/3Bpod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: bpod3
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: ctr
    image: skiibum/testpod:qos
    imagePullPolicy: IfNotPresent
    command: ["tail", "-f", "/dev/null"]
    resources:
      requests:
        cpu: "250m"
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .spec.containers[0].resources
{
  "requests": {
    "cpu": "250m"
  }
}
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .status.containerStatuses[0].resourcesAllocated
null
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# k get pod bpod3-127.0.0.1 -ojson | jq .status.containerStatuses[0].resources
null
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash#
```

  - Feature gate enabled, kubeletonly local cluster.
```bash
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /var/run/kubernetes/static-pods/3Bpod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: bpod3
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: ctr
    image: skiibum/testpod:qos
    imagePullPolicy: IfNotPresent
    command: ["tail", "-f", "/dev/null"]
    resources:
      requests:
        cpu: "250m"
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# 
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# crictl ps
CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID              POD
d06cf4835bc0a       254425ccdec9a       3 minutes ago       Running             ctr                 0                   d58670c526414       bpod3-127.0.0.1
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# 
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /sys/fs/cgroup/cpu/kubepods/burstable/pod07f9b33c5aa6780ad184c1db155963fb/d06cf4835bc0af6f54d13379aa1e2453e72387e3ace0cb2f973673915c258fc4/cpu.shares 
256
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# 
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /var/run/kubernetes/static-pods/3Bpod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: bpod3
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: ctr
    image: skiibum/testpod:qos
    imagePullPolicy: IfNotPresent
    command: ["tail", "-f", "/dev/null"]
    resources:
      requests:
        cpu: "125m"
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# crictl ps
CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID              POD
7890e9f519f0e       254425ccdec9a       15 seconds ago      Running             ctr                 0                   09250f693c922       bpod3-127.0.0.1
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /sys/fs/cgroup/cpu/kubepods/burstable/podd8dd8ced76f957ad395b2e579d3d3e2d/7890e9f519f0ec097c55dffd12eabb7696f8d620039561840d8bea972499b82d/cpu.shares
128
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash#
```

  - Feature gate disabled, kubeletonly local cluster.
```bash
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# 
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /var/run/kubernetes/static-pods/3Bpod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: bpod3
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: ctr
    image: skiibum/testpod:qos
    imagePullPolicy: IfNotPresent
    command: ["tail", "-f", "/dev/null"]
    resources:
      requests:
        cpu: "200m"
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# crictl ps
CONTAINER           IMAGE               CREATED              STATE               NAME                ATTEMPT             POD ID              POD
8f7a5e9000ed8       254425ccdec9a       About a minute ago   Running             ctr                 0                   0e20db4eb61f8       bpod3-127.0.0.1
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /sys/fs/cgroup/cpu/kubepods/burstable/pod0df8ebda9e77aef3325dcfe6985da39b/8f7a5e9000ed868b72a7c4dae93e82294007837b2d86856ca238a4e2de8d6f2c/cpu.shares 
204
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# 
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /var/run/kubernetes/static-pods/3Bpod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: bpod3
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: ctr
    image: skiibum/testpod:qos
    imagePullPolicy: IfNotPresent
    command: ["tail", "-f", "/dev/null"]
    resources:
      requests:
        cpu: "400m"
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# crictl ps
CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID              POD
ebf36203349d4       254425ccdec9a       30 seconds ago      Running             ctr                 0                   7aa3c12e76e05       bpod3-127.0.0.1
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# cat /sys/fs/cgroup/cpu/kubepods/burstable/pod2d1ad7bae07eb09ebb126229f381519d/ebf36203349d467d5cab2f22d4110b1cfad4aa4a9988a7a03a6e057c994fac44/cpu.shares 
409
root@vbuild:~/go/src/k8s.io/kubernetes-kubeletonly-crash# 
```